File: leetcode/110.balanced-binary-tree.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
IMBALANCED = -2


class Solution:
    def isBalanced(self, root: TreeNode | None) -> bool:
        return self.helper(root) != IMBALANCED

    def helper(self, node: TreeNode | None) -> int:
        if not node:
            return -1

        left_height = self.helper(node.left)

        if left_height == IMBALANCED:
            return IMBALANCED

        right_height = self.helper(node.right)

        if right_height == IMBALANCED or abs(left_height - right_height) > 1:
            return IMBALANCED

        return 1 + max(left_height, right_height)


# @leet end

File: leetcode/110.balanced-binary-tree.ts
// @leet start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function isBalanced(root: TreeNode | null): boolean {
  return getHeightAndHeightBalance(root)[1];
}

function getHeightAndHeightBalance(root: TreeNode | null): [number, boolean] {
  if (!root) {
    return [-1, true];
  }

  const [leftHeight, leftBalance] = getHeightAndHeightBalance(root.left);
  const [rightHeight, rightBalance] = getHeightAndHeightBalance(root.right);
  const balance =
    leftBalance && rightBalance && Math.abs(leftHeight - rightHeight) <= 1;

  return [1 + Math.max(leftHeight, rightHeight), balance];
}
// @leet end
